### DISCUSSION TITLE
How tauri implements a floating window.

### DISCUSSION THREAD

**Discussion by Lydanne** [tauri-apps/tauri#4452] [Q&A]
Recently, I want to use tauri to make a floating window application, but I have not found a way to create a floating window.

**Comment by xuchaoqian** [comment]
Maybe these apis can help you: 
[https://docs.rs/tauri/1.0.0/tauri/window/struct.Window.html#method.set_always_on_top](https://docs.rs/tauri/1.0.0/tauri/window/struct.Window.html#method.set_always_on_top)
[https://docs.rs/tauri/1.0.0/tauri/window/struct.WindowBuilder.html#method.decorations](https://docs.rs/tauri/1.0.0/tauri/window/struct.WindowBuilder.html#method.decorations)
[https://docs.rs/tauri/1.0.0/tauri/window/struct.WindowBuilder.html#method.transparent](https://docs.rs/tauri/1.0.0/tauri/window/struct.WindowBuilder.html#method.transparent)

**Reply by Lydanne** [reply]
Let me try.

**Reply by Lydanne** [reply]
![image](https://user-images.githubusercontent.com/39021696/175781289-f4ec7567-13bb-4c76-a554-2a021ddca725.png)
I can't overlay the window to the menu bar.

**Reply by Lydanne** [reply]
my code

```rust
#![cfg_attr(
    all(not(debug_assertions), target_os = "windows"),
    windows_subsystem = "windows"
)]

use tauri::{Manager, PhysicalPosition, Position};

fn main() {
    let context = tauri::generate_context!();

    tauri::Builder::default()
        .setup(|app| {
            let main_window = app.get_window("main").unwrap();
            main_window.set_always_on_top(true).unwrap();
            // main_window.set_fullscreen(true);
            main_window.set_decorations(false).unwrap();
            main_window
                .set_position(Position::Physical(PhysicalPosition { x: 0, y: -1000 }))
                .unwrap();

            Ok(())
        })
        .menu(tauri::Menu::os_default(&context.package_info().name))
        .run(context)
        .expect("error while running tauri application");
}

```

**Reply by Lydanne** [reply]
> [<img alt="" width="32" height="32" src="https://avatars.githubusercontent.com/u/1064052?s=64&amp;v=4"> xuchaoqian ](/xuchaoqian)

help

**Reply by xuchaoqian** [reply]
hi @FabianLars , is it possible to overlay the window to the menu bar?

> ![image](https://user-images.githubusercontent.com/39021696/175781289-f4ec7567-13bb-4c76-a554-2a021ddca725.png) I can't overlay the window to the menu bar.

**Reply by FabianLars** [reply]
@WumaCoder @xuchaoqian 
Yes but only by using the native apis manually. I don't have access to a mac so i can't test it, but it should go in this direction:
You need to add the `cocoa` crate for this to work
```rs
#![cfg_attr(
    all(not(debug_assertions), target_os = "windows"),
    windows_subsystem = "windows"
)]

use tauri::{Manager, PhysicalPosition, Position};

fn main() {
    let context = tauri::generate_context!();

    tauri::Builder::default()
        .setup(|app| {
            let main_window = app.get_window("main").unwrap();
            main_window.set_always_on_top(true).unwrap();
            // main_window.set_fullscreen(true);
            main_window.set_decorations(false).unwrap();
            main_window
                .set_position(Position::Physical(PhysicalPosition { x: 0, y: -1000 }))
                .unwrap();

            #[cfg(target_os = "macos")]
            {
              use cocoa::base::id;
              use cocoa::appkit::{NSWindow, NSMainMenuWindowLevel};
              let ns_win = main_window.ns_window().unwrap() as id;
              unsafe { ns_win.setLevel_((NSMainMenuWindowLevel + 1) as u64); }
            }

            Ok(())
        })
        .menu(tauri::Menu::os_default(&context.package_info().name))
        .run(context)
        .expect("error while running tauri application");
}
```

**Reply by Lydanne** [reply]
@FabianLars 
Unfortunately, this code still can't cover the menu bar, or can we change the idea, such as hiding the menu bar or cocoa `ns_win.setFloatingPanel(YES)`？
But the current version of cocoa does not support `setFloatingPanel`
